### PR TITLE
Add Search form helper

### LINKED ISSUE #27: Add a helper to generate search form
I wrote a helper that automagically creates a complete search form based on the table's `searchConfiguration()` fields. I think it would be useful to include in this plugin.

``` php
use Cake\View\Helper;
use Cake\ORM\TableRegistry;

/**
 * SearchForm helper
 */
class SearchFormHelper extends Helper
{

    public $helpers = ['Form', 'Html'];

    /**
     * Default configuration.
     *
     * @var array
     */
    protected $_defaultConfig = [];

    /**
     * Output a complete search form, based on the search configurations set up
     * in the specified table
     *
     * @param $model string The Table to load
     * @return null|string An formatted FORM with inputs.
     */
    public function form($model)
    {
        $table = TableRegistry::get($model);

        // make sure the table has the behavior and has implemented the searchConfiguration method
        if (
            ! $table->behaviors()->has('Search') ||
            ! method_exists($table, 'searchConfiguration')
        ) {
            return null;
        }

        $searchConfig = $table->searchConfiguration()->all();

        $output = $this->Html->tag('h3', __('Search'));
        $output .= $this->Form->create();

        foreach ($searchConfig as $object) {
            $config = $object->config();

            // you can add stuff like `'multiple' => true` or `'empty' => ' '`
            // to the individual searchConfiguration field calls to be passed to
            // the form input
            $output .= $this->Form->input($config['name'], $config);
        }

        $output .= $this->Form->submit();
        $output .= $this->Form->end();

        return $output;
    }
}
```

The helper is loaded into the view, as normal, and is called in the template like this

``` php
<?= $this->SearchForm->form('Articles') ?>
```

(or if you want to be lazy, and copy-paste the same thing in all of your templates, you can call it like this)

``` php
<?= $this->SearchForm->form($this->request->params['controller']) ?>
```

search-form

### PR DESCRIPTION
**Pull request by curtisgibby**
Fix #27

### CODE REVIEW

File: src/View/Helper/SearchFormHelper.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        if (
+            ! $table->behaviors()->has('Search') ||
+            ! method_exists($table, 'searchConfiguration')
+        ) {
+            return null;
+        }
+
+        $searchConfig = $table->searchConfiguration()->all();
+
+        $output = $this->Html->tag('h3', __('Search'));
```
Comment (ADmad) [inline src/View/Helper/SearchFormHelper.php]: Better to use `legend` tag of the form.

File: src/View/Helper/SearchFormHelper.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        foreach ($searchConfig as $object) {
+            $config = $object->config();
+
+            // you can add stuff like `'multiple' => true` or `'empty' => ' '`
+            // to the individual searchConfiguration field calls to be passed to
+            // the form input
+            $output .= $this->Form->input($config['name'], $config);
+        }
+
+        $output .= $this->Form->submit();
```
Comment (ADmad) [inline src/View/Helper/SearchFormHelper.php]: Using `button()` allows better styling.

**Comment by curtisgibby** [comment]
I've [updated](1a1cc341172df489c2aaf302b53a9946ecd9b8bf) the helper as suggested, using a form legend and a button.

File: src/View/Helper/SearchFormHelper.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        $fields = [];
+        foreach ($searchConfig as $object) {
+            $config = $object->config();
+
+            // you can add stuff like `'multiple' => true` or `'empty' => ' '`
+            // to the individual searchConfiguration field calls to be passed to
+            // the form input
+            $fields[$config['name']] = $config;
+        }
+        $output .= $this->Form->inputs($fields, ['legend' => __('Search')]);
```
Comment (ADmad) [inline src/View/Helper/SearchFormHelper.php]: Use domain `search` for translated strings

Comment (curtisgibby) [inline src/View/Helper/SearchFormHelper.php]: What about in the submit button's `__('Submit')`? There too?

**Comment by curtisgibby** [comment]
Is there anything else you'd want done before this PR can be merged?

File: src/View/Helper/SearchFormHelper.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @param string $model The Table to load
+     * @return null|string An formatted FORM with inputs.
+     */
+    public function form($model)
+    {
+        $table = TableRegistry::get($model);
+
+        // make sure the table has the behavior and has implemented the searchConfiguration method
+        if (! $table->behaviors()->has('Search') ||
```
Comment (dereuromark) [inline src/View/Helper/SearchFormHelper.php]: There should not be a space after `!`.

Comment (curtisgibby) [inline src/View/Helper/SearchFormHelper.php]: Interesting that it passed the Travis / PHP-CS as-is... but I'll make the change.

File: src/View/Helper/SearchFormHelper.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+use Cake\View\Helper;
+use Cake\View\View;
+
+/**
+ * SearchForm helper
+ */
+class SearchFormHelper extends Helper
+{
+
+    public $helpers = ['Form'];
```
Comment (ADmad) [inline src/View/Helper/SearchFormHelper.php]: Docblock missing.

**Comment by ADmad** [comment]
> Is there anything else you'd want done before this PR can be merged?

Tests :)

**Comment by curtisgibby** [comment]
I was afraid that you were going to say that. :)

The helper is dependent on the Table having the Behavior. How do you mock that in a test?

**Comment by ADmad** [comment]
Why mock it? Just attach the behavior to the table.

**Comment by curtisgibby** [comment]
I don't have an `Articles` table in my project; I've got tables that have nothing to do with your tables. My tables do have the behavior, so I could write tests that would pass for me, but they'd fail for you. That's what I'm asking how to do: how to test a Table having a behavior when I can't guarantee that the table will exist in the testing environment.

**Comment by ADmad** [comment]
Read the manual for how to use fixtures. The fixtures from are available to app/plugin tests too. So just create a table using one of the fixtures and attach behavior to that.

**Comment by burzum** [comment]
I like the idea but the way it is implemented it seems to be not very flexible. Good for prototyping but not good enough for production. There is not enough control about the fields, the order of them and their look and feel.

The helper could use a template for the form in which it renders the fields into and maybe a callback that can be used on each field to add stuff before/after each input field or alter it.

**Comment by ADmad** [comment]
Closing as this has been stalled for a long time. A new PR can be made taking into consideration all the feedback provided.

**Pull request closed without merging**